[rush-lib] Added dependency analysis and simulate functions to -rush build- and enriched logged telemetry data

hanksterhan opened the pull request:
Usage: ```
$ rush build --analysis -file telemetry_datestamp.json
$ rush build --simulate -file telemetry_datastamp.json
```

## Summary
When building projects, it is very helpful to have telemetry data to be able to look inside a build. I have enriched the logged telemetry data by adding the dependency graph and the individual project build time+status. In addition, I wrote two functions (analysis and simulate) that will process this enriched telemetry data to provide some useful dependency analysis. Analysis will identify the projects which are chokepoints during the build. Reducing the build times of these projects will directly reduce the overall build time by 1 second. Simulate will identify the optimal number of CPU cores to build your project.

- Both the ```--analysis``` and ```--simulate``` flags require an additional ```--file``` flag that will point to a telemetry file in the telemetry folder.

## How it was tested
Invoked ```rush build --analysis --file telemetry_datestamp.json``` and ```rush build --simulate --file telemetry_datestamp.json``` and the printed output was formatted correctly. The telemetry file was found to be enriched after runs of ```rush build``` or ```rush rebuild```. 

I implemented quite a few changes but I am very open to refactoring suggestions as this is not my code base. I did my best to limit the number of new imports and to reuse existing code. I apologize in advance that the initial code is in 1 commit - I started writing this code in a way that was not visible to the rushstack repository and had to transfer my work to a forked repository.I look forward to collaborating on this PR, thanks!

dmichon-msft commented on libraries/rush-lib/src/logic/operations/BuildTimePlugin.ts:
```diff
@@ -0,0 +1,83 @@
+    hooks.afterExecuteOperations.tap(
+      PLUGIN_NAME,
+      (result: IExecutionResult, context: ICreateOperationsContext): void => {
+        this._buildTimes = _setBuildTimes(result);
+      }
+    );
+    return this._buildTimes;
```
This is a little weird. This function will always return the same empty array. If you want to provide programmatic access to the time data, you would be best served providing a separate API to get the array.

hanksterhan replied:
I don't directly call this `apply` function for the BuildTimePlugin. I am able to directly call the _setBuildTimes() function and get the time data. How should I write the `apply` function in this case? Should I change it to a `void` return type and get rid of the return statement on line 39 or should I do a larger refactor where I don't implement `BuildTimePlugin` as a `IPhasedCommandPlugin`?

hanksterhan replied:
I simplified this function and wrote a separate API to get the array.

dmichon-msft commented on libraries/rush-lib/src/logic/operations/BuildTimePlugin.ts:
```diff
@@ -0,0 +1,83 @@
+        startTime,
+        endTime,
+        durationString,
+        name: operation.name!.split(' ')[0],
```
What is this attempting to extract from the operation name? If it's the project name, that can be had from `operation.associatedProject.packageName`.

hanksterhan replied:
the packageName is what I was trying to extract. Thanks for pointing me to this property

dmichon-msft commented on libraries/rush-lib/src/logic/operations/BuildTimePlugin.ts:
```diff
@@ -0,0 +1,83 @@
+
+    if (startTime && endTime) {
+      const { duration } = stopwatch;
+      const durationString: string = duration.toFixed(2);
```
Why convert to a string and then back into a number? Seems like it would make more sense to record the full number and only round at display time.

hanksterhan replied:
I borrowed code from ConsoleTimelinePlugin to do the time calculations and reused the `ITimelineRecord` interface. In hindsight I don't need that much structure in my interface, especially the string type for durationString. I will refactor this to be a smaller interface with a number as the type for duration. Good catch.

dmichon-msft commented on libraries/rush-lib/src/logic/operations/DependencyAnalysisPlugin.ts:
```diff
@@ -0,0 +1,330 @@
+
+export interface IDependencyGraphEntry {
+  dependencies: string[];
+  dependents: string[];
```
Because the edit distance between `dependency` and `dependent` is so small, I recommend using `consumer` as the opposite of `dependency`.

hanksterhan replied:
Good call. Thanks

dmichon-msft commented on libraries/rush-lib/src/logic/operations/DependencyAnalysisPlugin.ts:
```diff
@@ -0,0 +1,330 @@
+
+export function makeDependencyGraph(): IDependencyGraph {
+  const dependencyGraph: IDependencyGraph = {};
+  const rushConfiguration: RushConfiguration = RushConfiguration.loadFromDefaultLocation({
```
You can hand the plugin instance the `RushConfiguration` instance in its constructor. It is accessible in the context of the action.

hanksterhan replied:
Good call

dmichon-msft commented on libraries/rush-lib/src/logic/operations/DependencyAnalysisPlugin.ts:
```diff
@@ -0,0 +1,330 @@
+}
+
+export function _printDependencyAnalysis(terminal: ITerminal, result: IExecutionResult): void {
+  const dependencyGraph: IDependencyGraph = makeDependencyGraph();
```
The execution-time dependency graph is directly available by examining `result.operationResults.keys()`; the `Operation` instances all list their dependencies on `.dependencies` and consumers on `.consumers`.

hanksterhan replied:
This refactor makes the code simpler and cleaner. Thank you for pointing me to `result.operationResults`

dmichon-msft reviewed (changes requested):
There's a lot of duplicate work going on in this plugin to regenerate information that is already available in the plugin hook context.

dmichon-msft commented on libraries/rush-lib/src/cli/scriptActions/PhasedScriptAction.ts:
```diff
@@ -167,6 +179,25 @@ export class PhasedScriptAction extends BaseScriptAction<IPhasedCommandConfig> {
       new ConsoleTimelinePlugin(terminal).apply(this.hooks);
     }
+
+    const showDependencyAnalysis: boolean = this._dependencyAnalysisParameter
+      ? this._dependencyAnalysisParameter.value
+      : false;
+    if (showDependencyAnalysis) {
+      const telemetryFolderName: string = path.join(this.rushConfiguration.commonTempFolder, 'telemetry');
+      const fullPath: string = path.join(telemetryFolderName, this._telemetryFileParameter?.value ?? '');
+      dependencyAnalysis(fullPath, terminal);
+      return;
+    }
+
+    const doSimulation: boolean = this._simulateParameter ? this._simulateParameter.value : false;
+    if (doSimulation) {
+      const telemetryFolderName: string = path.join(this.rushConfiguration.commonTempFolder, 'telemetry');
+      const fullPath: string = path.join(telemetryFolderName, this._telemetryFileParameter?.value ?? '');
+      simulateBuildTime(fullPath, terminal);
+      return;
+    }
```
These would be better handled as a completely separate `rush analyze` or other action, since they don't depend on any of the context of the build itself and are purely post-build static analysis operations.

hanksterhan replied:
Good call. I was able to add an Analyze action triggered via `rush analyze`

hanksterhan commented:
- New Rush Action `rush analyze` implemented instead of piggybacking on `rush build`
- Implemented `rush analyze --dependency-analysis` and `rush analyze --simulate` 
- `makeDependencyGraph` creates the dependencyGraph from the results as opposed to creating it from the RushConfiguration.
- Refactored `dependents` to `consumers`

octogonz commented on libraries/rush-lib/src/cli/actions/AnalyzeAction.ts:
```diff
@@ -0,0 +1,67 @@
+        'to build your project to help teams make more informed decisions about the number of CPU cores to use to build their project.'
+    ];
+    super({
+      actionName: 'analyze',
```
Should we give this a more specific name such as "rush analyze-telemetry"?

dmichon-msft replied:
That would probably make sense unless we want this function to be a dispatcher to a variety of different analysis commands. Unfortunately, `@rushstack/ts-command-line` doesn't handle the latter scenario very well, so it should probably be a more specific name.

hanksterhan replied:
Updated the cli call to `rush analyze-telemetry`

octogonz commented on common/changes/@microsoft/rush/dependency-analysis_2022-08-16-16-30.json:
```diff
@@ -0,0 +1,10 @@
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Enriched the logged telemetry data by appending the build machine’s information e.g. number of cores, architecture, etc., individual project build times and status, and the dependency graph for the projects. An additional command line action `rush analyze` was added to do analysis on the enriched telemetry data. `rush analyze —dependency-analysis` and `rush analyze —simulate` we ...
```
```suggestion
      "comment": "Improved the logged telemetry data to include more details such as number of CPU cores, architecture, individual project build times and status, and the dependency graph for the projects",
```

This part should probably be a separate `rush change` entry:

> Added a new command "rush analyze-telemetry" which can identify projects with the largest impact on build times, as well as calculating the ideal amount of parallelism

The CHANGELOG.md entries are quick summaries, so we don't need to include too many details

octogonz commented on common/reviews/api/rush-lib.api.md:
```diff
@@ -127,6 +127,9 @@ export class CredentialCache {
 }
 
+// @beta (undocumented)
+export function dependencyAnalysis(filename: string, terminal: ITerminal): void;
```
Our coding convention is to group APIs into meaningful containers such as a class.  We try to avoid exporting loose functions. (If `dependencyAnalysis()` is a top-level item in a large .api.md file, what "dependency" does it mean? What feature is it related to? It is not obvious.)  Also, the function name should generally start with a verb.

Consider an API design like this:

```ts
/**
 * Exposes the underlying operations of the "rush analyze-telemetry" command.
 */
export class TelemetryAnalysis {
  public static printProjectBuildTimeReport(telemetryFilePath: string, terminal: ITerminal): void;
  public static printCpuCoresReport(telemetryFilePath: string, terminal: ITerminal): void;
}
```

I am also curious how you are using this API.  The caller could simply invoke the `rush analyze-telemetry` shell command -- what is the advantage of calling the TypeScript API?

hanksterhan replied:
The API design you recommend look great. I will work on this refactoring. Per your last point - are you recommending that the telemetry analysis functions in DependencyAnalysisPlugin.ts could live in AnalyzeAction.ts?

dmichon-msft replied:
This is relevant specifically to the public API surface more than internal usage. If the API isn't exported from the index file (which it may not need to be), use whatever shape is most ergonomic for unit testing and execution.

octogonz commented on common/reviews/api/rush-lib.api.md:
```diff
@@ -220,6 +223,16 @@ export class FileSystemBuildCacheProvider {
 export type GetCacheEntryIdFunction = (options: IGenerateCacheEntryIdOptions) => string;
 
+// @beta (undocumented)
+export interface IBuildTimeRecord {
```
To clarify that these interfaces are related to telemetry, maybe we could add prefixes:

- `IBuildTimeRecord` --> `ITelemetryBuildTimeRecord`
- `IDependencyGraph` --> `ITelemetryDependencyGraph`
- `IDependencyGraphEntry` --> `ITelemetryDependencyGraphEntry`
- `IMachineInfo` --> `ITelemetryMachineInfo`
- . . .

octogonz commented on common/reviews/api/rush-lib.api.md:
```diff
@@ -281,6 +294,11 @@ export interface ICredentialCacheOptions {
     supportEditing: boolean;
 }
 
+// Warning: (ae-forgotten-export) The symbol "IDependencyGraphEntry" needs to be exported by the entry point index.d.ts
```
Export the missing declaration

hanksterhan replied:
Removed altogether

octogonz commented on common/reviews/api/rush-lib.api.md:
```diff
@@ -349,6 +367,22 @@ export interface ILogger {
+// @beta (undocumented)
+export interface IMachineInfo {
+    // (undocumented)
+    machineArch: string;
```
```suggestion
    machineArchitecture: string;
```

octogonz commented on libraries/rush-lib/src/cli/scriptActions/PhasedScriptAction.ts:
```diff
@@ -541,6 +545,17 @@ export class PhasedScriptAction extends BaseScriptAction<IPhasedCommandConfig> {
+        machineCPU: os.cpus()[0].model,
+        machineCores: os.cpus().length,
+        machineTotalMemMB: Math.floor(os.totalmem() / 1024 / 1024),
+        machineFreeMemMB: Math.floor(os.freemem() / 1024 / 1024)
```
Would `Math.round()` be more appropriate?

octogonz commented on common/reviews/api/rush-lib.api.md:
```diff
@@ -904,6 +944,14 @@ export class RushUserConfiguration {
+// Warning: (ae-incompatible-release-tags) The symbol "_setBuildTimes" is marked as @beta, but its signature references "IExecutionResult" which is marked as @alpha
+//
+// @beta (undocumented)
+export function _setBuildTimes(result: IExecutionResult): IBuildTimeRecord[];
```
Why does this API need to be exported at all?

hanksterhan replied:
So that the telemetry logger in `PhasedScriptAction` can get the Build Timings in a formatted way. Would it be better to rewrite the `_setBuildTimes()` function inside of `PhasedScriptAction`?

dmichon-msft replied:
This needs to be exported from the internal module, but doesn't need to be exported from the index file.

hanksterhan replied:
I see what you mean now. I think most if not all of the functions I exported in the index file can be taken back. I will go through them

octogonz commented on common/reviews/api/rush-lib.api.md:
```diff
@@ -904,6 +944,14 @@ export class RushUserConfiguration {
     static initializeAsync(): Promise<RushUserConfiguration>;
 }
 
+// Warning: (ae-incompatible-release-tags) The symbol "_setBuildTimes" is marked as @beta, but its signature references "IExecutionResult" which is marked as @alpha
```
To fix this warning, you could just promote `IExecutionResult` to be `@beta`.  The `@alpha` vs `@beta` distinction is not very meaningful for `rush-lib`

dmichon-msft commented on libraries/rush-lib/src/index.ts:
```diff
@@ -84,6 +84,12 @@ export { IOperationRunner, IOperationRunnerContext } from './logic/operations/IO
+export { IBuildTimeRecord, _setBuildTimes } from './logic/operations/BuildTimePlugin';
+export {
+  IDependencyGraph,
+  dependencyAnalysis,
+  simulateBuildTime
```
Do these functions need to be part of the public API of rush-lib?

hanksterhan replied:
No they do not - I removed them.

dmichon-msft commented on libraries/rush-lib/src/cli/scriptActions/PhasedScriptAction.ts:
```diff
@@ -541,6 +545,17 @@ export class PhasedScriptAction extends BaseScriptAction<IPhasedCommandConfig> {
         countNoOp: 0
       };
 
+      let buildTimings: IBuildTimeRecord[] = [];
```
Could you move all the logic in the `if (this.parser.telemetry)` to be the implementation of a single `TelemetryPlugin` or similar that also includes all the logic for recording the graph and the build times?

dmichon-msft commented on libraries/rush-lib/src/cli/scriptActions/PhasedScriptAction.ts:
```diff
@@ -576,13 +591,18 @@ export class PhasedScriptAction extends BaseScriptAction<IPhasedCommandConfig> {
-        extraData
+        extraData,
+        machineInfo,
+        buildTimings,
+        dependencyGraph
```
Once we're including this data in the telemetry file, the summary fields above become redundant and we may as well just list out the full dependency graph, with the timing and status for each operation in the graph.

dmichon-msft commented on libraries/rush-lib/src/logic/Telemetry.ts:
```diff
@@ -41,6 +55,9 @@ export interface ITelemetryData {
   readonly rushVersion?: string;
   readonly extraData?: { [key: string]: string | number | boolean };
+  readonly machineInfo?: IMachineInfo;
+  readonly buildTimings?: IBuildTimeRecord[];
+  readonly dependencyGraph?: IDependencyGraph;
```
Why are these separate objects? We could just list the timings and status values on the graph nodes.

dmichon-msft commented on libraries/rush-lib/src/logic/operations/BuildTimePlugin.ts:
```diff
@@ -0,0 +1,85 @@
+      data.push({
+        startTime,
+        duration,
+        name: operation.associatedProject?.packageName ?? '',
```
Using the package name instead of the operation name artificially limits the utility of this feature. Test it out on running `rush retest` within the rushstack repository and you will see the problem immediately.

dmichon-msft commented on libraries/rush-lib/src/logic/operations/DependencyAnalysisPlugin.ts:
```diff
@@ -0,0 +1,296 @@
+  buildTime: number;
+}
+
+export class DependencyAnalysisPlugin implements IPhasedCommandPlugin {
```
With the way you have currently integrated with PhasedScriptAction, this doesn't make sense as a plugin at present. I would prefer if all of the telemetry code was migrated into a single plugin that uses the afterExecuteOperations hook and owns the call to `telemetry.log`.

dmichon-msft reviewed (dismissed):
Project vs. operation is an important distinction. Phased builds work in terms of operations; projects are merely a tool for helping define operations, so we should be doing all analysis in operation space.

dmichon-msft commented:
I'm going to create a separate PR to add the requested telemetry data to the default telemetry files. I believe it would be cleanest to then convert the analyzer to a standalone CLI tool (i.e. its own package, not part of rush-lib) that can operate on a telemetry file in the standard format. This will extend the utility of the feature to also work with the new multi-phase Heft work that is currently ongoing, allow faster iteration on the tool itself, and not bind it to the guts of rush-lib.

octogonz commented:
> I'm going to create a separate PR to add the requested telemetry data to the default telemetry files. I believe it would be cleanest to then convert the analyzer to a standalone CLI tool (i.e. its own package, not part of rush-lib) that can operate on a telemetry file in the standard format. This will extend the utility of the feature to also work with the new multi-phase Heft work that is currently ongoing, allow faster iteration on the tool itself, and not bind it to the guts of rush-lib.

I agree that this will reduce the amount of API reviews that would slow down this feature.

That said, I think it's a cool feature for Rush that we would be interested to showcase and help maintain.

- If you want to create a separate CLI tool, we could still host it in the https://github.com/microsoft/rushstack repo (e.g. under the `apps/analyze-telemetry` folder, maybe published as `@rushstack/analyze-telemetry`?).  

- Another idea would be to keep it integrated with Rush as a `rush analyze-telemetry` command, but implement it as a [plugin package](https://rushjs.io/pages/extensibility/creating_plugins/).  There's a gallery of example projects [here](https://github.com/bytesfriends/rush-plugins/tree/main/rush-plugins), and a CLI example here:  [rush-sort-package-json/command-line.json](https://github.com/bytesfriends/rush-plugins/blob/main/rush-plugins/rush-sort-package-json/command-line.json).  If you go this route, then your project folder could be something like `rush-plugins/rush-analyze-telemetry-plugin`.

@hanksterhan BTW sorry if there seemed to be a fire hose of feedback for this PR.  The ideal flow is something like:

`prototype` --> `design discussion` --> `design note` --> `pull request` --> `code review`

For open source projects, these steps often happen in a different order. 😄

hanksterhan commented:
@octogonz @dmichon-msft Thank you for all the feedback. I will work on the changes but I will wait to lean on the telemetry files created as part of #3596

hanksterhan commented on libraries/rush-lib/src/logic/operations/DependencyAnalysisPlugin.ts:
```diff
@@ -0,0 +1,284 @@
+    const consumers: string[] = [];
+
+    operation.dependencies.forEach((dependency: Operation) => {
+      dependencies.push(dependency.associatedProject?.packageName ?? '');
```
This was handled by the split PR #3596

hanksterhan commented:
I have refactored all the analysis to be done inside of AnalyzeTelemetryAction instead of in BuildTimePlugin, DependencyAnalysisPlugin, and PhasedActionScript. The next step is to follow octogonz's direction and implement rush `analyze-telemetry` as a plugin package
